client/src/reducers/alert.js

```diff
@@ -0,0 +1,16 @@
+import { alert } from '../sagas/types';
+
+const actions = alert.reducer;
+
+const initialState = [];
+
+export default (state = initialState, { type, payload }) => {
+  switch (type) {
+    case actions.ADD_ALERT:
+      return [payload, ...state];
+    case actions.REMOVE_ALERT:
+      return state.filter(a => a.id !== payload);
```

Comment:
I've been tryna fumble around with whatever the heck `payload`/`state` are here, but I've never been able to nail down exactly what's happening.

1. Is `{ type, paylod }` a map with two elements?
1. What is `type`? (obviously we're switching on it, and we expect it to be one of the two alert actions, otherwise we just return `state`)
1. what does `...state` do  exactly? is that just inserting all of the elements of `state` into the array that will be the state elements PLUS `payload`?
1. What exactly is js doing with `!==` on `a.id !== payload`?

Reply:
1. `{ type, payload }` is [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) the input object (which represents the action triggered by the saga) into the `type` and `payload` vars. It's essentially just syntactic sugar in ES6.
2. `type` is the type of action being triggered
3. `...` is the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) in JS. Here, we're prepending `payload` to the existing `state`, which is an array
4. JS fun :) `!==` is not equivalent to `!=`. The former compares type _and_ value; the latter compares value only. For example: `5 == "5"` returns true (:oof:), but `5 === "5"` returns false. [More details](https://www.w3schools.com/js/js_comparisons.asp)